src/piton/jax.py

```diff
@@ -0,0 +1,1095 @@
+
+        res = xla_client.ops.CustomCallWithLayout(
+            ctx.builder,
+            name.encode("utf8"),
+            operands=[a, b, indices, g],
+            shape_with_layout=xla_client.Shape.tuple_shape(
+                [
+                    _convert_to_xla_shape(
+                        val, force_dtype=xla_client.dtype_to_etype(np.float32)
+                    )
+                    for val in (da_shape, db_shape)
+                ]
+            ),
+            operand_shapes_with_layout=[
+                _convert_to_xla_shape(val)
+                for val in (a_shape, b_shape, indices_shape, g_shape)
+            ],
+            opaque=struct.pack(
+                "IIII",
+                a_shape.shape[1],
+                indices_shape.shape[0],
+                a_shape.shape[0],
+                b_shape.shape[0],
+            ),
+        )
+
+        return [
+            _convert_back(xla_client.ops.GetTupleElement(res, i))
+            for i, _ in enumerate(avals_out)
+        ]
+
+    raise ValueError(
+        "Unsupported platform; this must be either 'cpu' or 'gpu', got "
+        + str(ctx.platform)
+    )
+
+
+embedding_dot_backward_p.def_abstract_eval(
+    embedding_dot_backward_p_abstract_eval
+)
+embedding_dot_backward_p.multiple_results = True
+embedding_dot_backward_p.def_impl(grad(embedding_dot_fallback, argnums=(0, 1)))
+xla.register_translation(
+    embedding_dot_backward_p, embedding_dot_backward_p_xla_translation
+)
+
+
+def exp_mean_fallback(a, b, c):
+    matrix = jnp.matmul(a, b.T)
+    return jnp.exp2(matrix + c).astype(jnp.float32).mean().astype(a.dtype)
+
+
+@partial(custom_vjp, nondiff_argnums=(2,))
+def exp_mean(a, b, sparse_c):
+    return exp_mean_fwd(a, b, sparse_c)[0]
+
+
+@jax.jit
+def exp_mean_fwd(a, b, sparse_c):
+    r, da, db = exp_mean_p.bind(a, b, *sparse_c)
+    return r, (a, da, db)
+
+
+@jax.jit
+def exp_mean_bwd(_sparse_c, res, g):
+    (
+        a,
+        da,
+        db,
+    ) = res
+    # Make sure we can safely do the multiplication
+    assert da.dtype == g.dtype
+
+    return ((da * g).astype(a.dtype), (db * g).astype(a.dtype))
+
+
+exp_mean.defvjp(exp_mean_fwd, exp_mean_bwd)
+
+exp_mean_p = core.Primitive("exp_mean")
+
+
+def exp_mean_abstract_eval(a, b, offsets, defaults, indices, values):
+    assert a.dtype == b.dtype
+    assert len(a.shape) == len(b.shape) == 2
+    assert (
+        len(offsets.shape)
+        == len(defaults.shape)
+        == len(indices.shape)
+        == len(values.shape)
+        == 1
+    )
+
+    assert a.shape[1] == b.shape[1]
+
+    assert offsets.shape == (a.shape[0] + 1,)
+    assert defaults.shape == (a.shape[0],)
+
+    assert indices.shape == values.shape
+    assert len(indices.shape) == 1
+
+    assert a.dtype == b.dtype
+    assert offsets.dtype == jnp.uint32
+    assert defaults.dtype == values.dtype == jnp.float32
+    assert indices.dtype == jnp.uint32
+
+    assert a.shape[0] % 32 == 0
```

Comment:
The user is expected to design their weight matrics so that everything is divisible by 32.

We can remove this requirement in the future by careful CUDA tweaks, but this is the simplest solution for now.